File: src/Transaction.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

/**
 * Transaction class handles the processing of transactions loaded from the daily 
 * transaction log
 *
 */

public class Transaction {
	/**
	 * Transaction class constructor
	 * 
	 * @param acct
	 * @param avail_tickets
	 */
	
	public Transaction(Accounts acct, AvailableTickets avail_tickets){
		this.filereader(acct, avail_tickets);
	}
	
	/**
	 * create transaction processes the transaction that indicates that an account was created
	 * 
	 * @param acct
	 * @param avail_tickets
	 * @param transaction_line
	 */
	public void createTransaction(Accounts acct, AvailableTickets avail_tickets, String transaction_line[]){
		acct.create(transaction_line[1], transaction_line[2], Double.parseDouble(transaction_line[3]));	// creates account
	}
	/**
	 * deteteTransaction processes the transaction that indicates that an account was deleted
	 * 
	 * @param acct
	 * @param avail_tickets
	 * @param transaction_line
	 */
	public void deleteTransaction(Accounts acct, AvailableTickets avail_tickets, String transaction_line[]){
		acct.delete(transaction_line[1]);	// deletes account
	}
	
	/**
	 * sellTransaction processes the transaction that indicates that a seller wants to sell tickets
	 * 
	 * @param acct
	 * @param avail_tickets
	 * @param transaction_line
	 */
	public void sellTransaction(Accounts acct, AvailableTickets avail_tickets, String transaction_line[]){
		Ticket ticket = new Ticket(transaction_line[1], transaction_line[2], Double.parseDouble(transaction_line[4]));	//create a new ticket object from the line
		avail_tickets.create(ticket, Integer.parseInt(transaction_line[3]));	//adds the ticket
	}
	
	/**
	 * buyTransaction processes the transaction that indicates that a buyer wants to buy tickets
	 * 
	 * @param acct
	 * @param avail_tickets
	 * @param transaction_line
	 */
	public void buyTransaction(Accounts acct, AvailableTickets avail_tickets, String transaction_line[]){
		Ticket ticket = new Ticket(transaction_line[1], transaction_line[2], Double.parseDouble(transaction_line[4]));	//create a new ticket object from the line
		Integer previous_ticket_count = avail_tickets.getAvailableTickets(ticket);	//get the amount of tickets available before the sale
		if ((previous_ticket_count - Integer.parseInt(transaction_line[3]) < 0)) {	//check to see that the transaction won't result in a negetive ticket amount
			System.out.println("ERROR: Cannot buy more than the available amount of tickets. Available:" + previous_ticket_count + " Wanted to buy:" + transaction_line[3] + " Transaction:" + transaction_line[0] + "_" +transaction_line[1] + "_" +transaction_line[2] + "_" +transaction_line[3] + "_" +transaction_line[4]); //throw error if it does
		}
		else {
			avail_tickets.setAvailableTickets(ticket, previous_ticket_count - Integer.parseInt(transaction_line[3])); //changes the amount of tickets available to reflect the sale
		}
		
	}
	
	/**
	 * refundTransaction processes the transaction log that indicates a refund was performed
	 * 
	 * @param acct
	 * @param avail_tickets
	 * @param transaction_line
	 */
	public void refundTransaction(Accounts acct, AvailableTickets avail_tickets, String transaction_line[]){
		Double buyer_previous_credit = acct.getAvailableCredit(transaction_line[1]);	// gets previous buyer credit
		Double seller_previous_credit = acct.getAvailableCredit(transaction_line[2]);	// gets previous seller credit
		acct.setAvailableCredit(transaction_line[1], buyer_previous_credit + Double.parseDouble(transaction_line[3]));	// sets new buyer credit
		acct.setAvailableCredit(transaction_line[2], seller_previous_credit - Double.parseDouble(transaction_line[3]));	// sets new seller credit
	}
	/**
	 * addCreditTransaction processes the transaction log that indicates credit was added
	 * 
	 * @param acct
	 * @param avail_tickets
	 * @param transaction_line
	 */	
	public void addCreditTransaction(Accounts acct, AvailableTickets avail_tickets, String transaction_line[]){
		Double previous_credit = acct.getAvailableCredit(transaction_line[1]);	// gets previous credit
		acct.setAvailableCredit(transaction_line[1], previous_credit + Double.parseDouble(transaction_line[3]));	// sets new credit
	}
	
	/**
	 * filereader reads the transaction log file and perform the appropriate action to 
	 * process each transaction
	 * 
	 * @param acct
	 * @param avail_tickets
	 */
	public void filereader(Accounts acct, AvailableTickets avail_tickets){
		String transaction_file = "transaction.txt";	// sets the transaction files location
		BufferedReader br = null;
		String line = "";
		String Delimiter = "_";		// sets the delimiter to "_"
	 
		try {
	 
			br = new BufferedReader(new FileReader(transaction_file));
			while ((line = br.readLine()) != null) {	// reads each line one at a time until null
		 		String[] transaction_line = line.split(Delimiter);	// splits the line on the delimiter
		 		
				switch (Integer.parseInt(transaction_line[0])) {	// checks the transaction code
				
				case 1: 
					if (transaction_line[0].isEmpty() || transaction_line[1].isEmpty() || transaction_line[2].isEmpty() || transaction_line[3].isEmpty())	//check for bad inputs
						System.out.print("ERROR: Bad input, missing fields. File:" + transaction_file);
					else {
						this.createTransaction(acct, avail_tickets, transaction_line);	// if code is 1, create is performed
					}
					break;
					
				case 2: 
					if (transaction_line[0].isEmpty() || transaction_line[1].isEmpty() || transaction_line[2].isEmpty() || transaction_line[3].isEmpty())	//check for bad inputs
						System.out.print("ERROR: Bad input, missing fields. File:" + transaction_file); //Throw error with discription and offending file name
					else {
						this.deleteTransaction(acct, avail_tickets, transaction_line);	// if code is 2, delete is performed
					}
					break;
				
				case 3: 
					if (transaction_line[0].isEmpty() || transaction_line[1].isEmpty() || transaction_line[2].isEmpty() || transaction_line[3].isEmpty() || transaction_line[4].isEmpty())	//check for bad inputs
						System.out.print("ERROR: Bad input, missing fields. File:" + transaction_file); //Throw error with discription and offending file name
					else {
						sellTransaction(acct, avail_tickets, transaction_line);
					}
					break;
				
				case 4: 
					if (transaction_line[0].isEmpty() || transaction_line[1].isEmpty() || transaction_line[2].isEmpty() || transaction_line[3].isEmpty() || transaction_line[4].isEmpty())	//check for bad inputs
						System.out.print("ERROR: Bad input, missing fields. File:" + transaction_file); //Throw error with discription and offending file name
					else {
						buyTransaction(acct, avail_tickets, transaction_line);
					}
					break;
					
				case 5: 	// if code is 5, refund is performed
					if (transaction_line[0].isEmpty() || transaction_line[1].isEmpty() || transaction_line[2].isEmpty() || transaction_line[3].isEmpty())	//check for bad inputs
						System.out.print("ERROR: Bad input, missing fields. File:" + transaction_file); //Throw error with discription and offending file name
					else {
						this.refundTransaction(acct, avail_tickets, transaction_line);
					}
					break;
				
				case 6: 	// if code is 6, addcredit is performed
					if (transaction_line[0].isEmpty() || transaction_line[1].isEmpty() || transaction_line[2].isEmpty() || transaction_line[3].isEmpty())	//check for bad inputs
						System.out.print("ERROR: Bad input, missing fields. File:" + transaction_file); //Throw error with discription and offending file name
					else {
						this.addCreditTransaction(acct, avail_tickets, transaction_line);
					}
					break;
				
				
	
				default:	
					System.out.print("ERROR: Bad input, incorrect transaction code: " + transaction_file); //Throw error with discription and offending file name
					break;
				}
					
		 
			}
	 
		} catch (FileNotFoundException e) {
			e.printStackTrace();	// prints File Not Found errors
		} catch (IOException e) {
			e.printStackTrace();	// prints IO errors
		} finally {
			if (br != null) {
				try {
					br.close();		// closes the buffered reader

				} catch (IOException e) {
					e.printStackTrace();	// print IO errors
				}
			}
		}
	}
	
	
	
}

File: src/transaction.cpp
#include "transaction.h"
#include "Accounts.h"
#include "StringBuilder.h"
#include <string.h>
using namespace std;

//static const string Daily = "/Transaction/Daily/daily.trans";
static const string Daily = "daily.trans";
static const string Users = "/Transaction/Users/current.trans";
static const string Tickets = "/Transaction/Tickets/avail.trans";
stringBuilder sb;

Transaction::Transaction(Accounts& account): acct(account){
	
}

/**
 * create method creates a user account and takes user name and account 
 * type
 */
void Transaction::create(string user_name, string user_type, double credit){
	if(acct.acc_type.find(user_name)==acct.acc_type.end()){
		acct.acc_type[user_name]=user_type;
		acct.avail_credit[user_name]= credit;
		string transaction = sb.Buildstring(1,user_name,user_type,credit);
		writeTransaction(transaction);
		cout << "Account Successfully Created\n";
	}else{
		cout << "Account Already Exists\n";
	}

}

/**
 * writeTransaction method writes the transaction string to the appropriate 
 * transaction file. 
 */
void Transaction::writeTransaction(string transaction){
	
	std::ofstream ofs;
	ofs.open (Daily, std::ofstream::out | std::ofstream::app);
  	ofs << transaction << endl;
  	ofs.close();

}
/**
 * Refund method processes the refund from a sellers account to the buyers account
 *
 */
void Transaction::refund(string buyer, string seller, double amount){
	if(acct.isAdmin()){
		if(acct.acc_type.find(buyer)!=acct.acc_type.end()){
			if(acct.acc_type.find(seller)!=acct.acc_type.end()){
				acct.avail_credit[buyer]+=amount;
				acct.avail_credit[seller]-=amount;
				string transaction = sb.BuildstringRefund(5,buyer,seller,amount);
				writeTransaction(transaction);
			}else{
				cout << "Invalid Seller Name\n";
			}
		}else{
			cout << "Invalid Buyer Name\n";
		}
	}else{
		cout << "Insufficient Privileges\n";
	}
				

}
/**
 * Addcredit method adds credit to a specified users account
 *
 */
void Transaction::addcredit(string user_name, double credit){
	if(acct.isBuyer()){
		if((acct.avail_credit[user_name]+credit)<999999.00){
			acct.avail_credit[user_name]+=credit;
			cout << "Credit Added Successfully\n";
			string transaction = sb.Buildstring(06, user_name, acct.acc_type[user_name], credit);
			writeTransaction(transaction);
		}else{
			cout << "Credit Limit Exceeded\n";
		}
	}else{
		cout << "Insufficient Privileges\n";
	}
}
/**
 * Sell method adds an event and assigns it a specified number of tickets
 * and a specified price.
 *
 */
void Transaction::sell(string event_name, string seller, int tickets, double price){
	string transaction = sb.Buildstring(3, event_name, seller, tickets, price);
	writeTransaction(transaction);
}

void Transaction::buy(string event_name, int tickets, string seller){
	if ((acct.num_tickets[event_name+seller] - tickets) < 0 )
	{
		cout << "Error, Cannot buy more tickets than than available. Wanted:" << tickets << " Avaialable:" << acct.num_tickets[event_name+seller] << "\n";
	} 
	else if ((tickets * acct.price_tickets[event_name+seller]) > acct.avail_credit[acct.current_user]){
		cout << "Error, User does not have enough credit to buy tickets. Cost:" << (tickets * acct.price_tickets[event_name+seller]) << " Credits:" <<acct.avail_credit[acct.current_user] << "\n" ; 
	}
	else
	{
		acct.num_tickets[event_name+seller] = acct.num_tickets[event_name+seller] - tickets;
		acct.avail_credit[acct.current_user] = acct.avail_credit[acct.current_user] - (tickets * acct.price_tickets[event_name+seller]);
		string transaction = sb.Buildstring(4, event_name, seller, tickets, acct.price_tickets[event_name+seller]);
		writeTransaction(transaction);
		transaction = sb.BuildstringRefund(5,seller,acct.current_user,(tickets * acct.price_tickets[event_name+seller]));
		writeTransaction(transaction);
	}
	
				
}
/**
 * Remove method deletes a users account and all tickets and events
 * associated with that account.
 *
 */
 
void Transaction::remove(string user_name){
	if(user_name.compare(acct.current_user)==0){
		cout << "Cannot remove own account\n";
	}else if(acct.acc_type.find(user_name)!=acct.acc_type.end()){
		//remove from available tickets log as well
		double credit = acct.avail_credit[user_name];
		string user_type = acct.acc_type[user_name];
		string transaction = sb.Buildstring(2,user_name,user_type,credit);
		acct.remove(user_name);
		writeTransaction(transaction);
	}else{
		cout << "Account Does Not Exist\n";
	}
}

/**
 * login method checks to see if the user account exists and then
 * assigns the account to be the current user
 */
void Transaction::login(string user_name){
	acct.current_user = user_name;
	if(acct.isValidAccount(acct.current_user)){
		acct.logged_in = true;
		cout << "Login Successful\n";
	
	}else{
		acct.current_user = "";
		acct.logged_in = false;
		cout << "Invalid Account\n";
	}
}
/**
 * logout method checks if there is a user logged in and if they are,
 * they get logged out.
 */
void Transaction::logout(){
	//write to transaction file
	string transaction = sb.Buildstring(0,acct.current_user, acct.acc_type[acct.current_user], acct.avail_credit[acct.current_user]);
	writeTransaction(transaction);
	acct.current_user = "";
	acct.logged_in = false;
	cout << "Logout Successful\n";
	
}

